### PR TITLE
[diffusion, cfg, tests] feat: add shared remove-padding interface

### PR DESCRIPTION
**Pull request by NancyFyong**
### What does this PR do?

Add a shared, opt-in **diffusion Actor remove-padding interface** using verl's existing configuration name:

```bash
actor_rollout_ref.model.use_remove_padding=true
```

The diffusion default is **false**. This PR supplies the configuration, capability checks, engine hooks, shared sequence-boundary/self-attention utility and CPU contracts. It does **not** enable a built-in model adapter: unsupported adapters fail before loading weights rather than silently ignoring the option.

**Independent main-based scope:** this branch is based directly on `main` (`8d2a9fd5`), not on #592. It carries no H3 adapter, rollout, recipe or determinism changes from that PR. #592 retains the model-specific H3 implementation and can later rebase onto this interface. This PR does not depend on #592 or claim its GPU results.

### Checklist Before Starting

- [x] Search for similar PRs: [diffusion packing](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+diffusion+packing), [use_remove_padding](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+use_remove_padding), and [396 in:body](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+396+in%3Abody). Read #396 and the existing #592. #592 owns H3 packed forward, while #562 concerns rollout request batching; neither supplies this common opt-in engine/adapter contract. The split is recorded on #396 for non-duplication context; this PR does not close that issue.
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Test

Validated on this main-based split, not only on the earlier H3 worktree:

```bash
export PYTHONPATH="$PWD"
export CUDA_VISIBLE_DEVICES=''
export TORCH_COMPILE_DISABLE=1 TORCHINDUCTOR_DISABLE=1 OMP_NUM_THREADS=2
python -m pytest -q \
  tests/workers/test_diffusion_packing_on_cpu.py \
  tests/utils/test_diffusion_packing_on_cpu.py \
  $(find tests -type f \( \
    -name '*minimax*on_cpu.py' -o \
    -name '*diffusers*on_cpu.py' -o \
    -name '*diffusion_attention*on_cpu.py' -o \
    -name 'test_diffusion_config_on_cpu.py' \
  \) -print) \
  tests/workers/rollout/rollout_vllm/test_vllm_omni_strategy_on_cpu.py \
  tests/special_sanity --disable-warnings
```

**Result: 363 passed, 20 warnings.** The 21 new cases cover configuration/defaults, unchanged Omni configuration, both custom and AutoModel load paths, FSDP/FSDP2 dispatch, unsupported adapter/engine/SP rejection, disabled-path preservation, native attention outputs and gradients, sample isolation, FA3 argument boundaries and fail-closed kernel loading. FA3 is mocked in CPU tests; this is not GPU-kernel evidence.

All tracked files passed all 12 pre-commit hooks, including generated YAML verification:

```bash
git ls-files -z | xargs -0 pre-commit run --show-diff-on-failure --files
git diff --cached --check
```

`pre-commit run --all-files` could not enumerate files because the host Git lacks `git ls-files --deduplicate`; the explicit tracked-file invocation above completed successfully. Generated configs were regenerated with `bash scripts/generate_trainer_config.sh`, staged, and then verified by the hook.

Runtime dependency provenance matches the branch pins: verl `fefb080262e1c015a0ea05f958822a6a512dc795`, vLLM-Omni `ded8934626aaad1a3e816c3a1d9d742efc012d93`.

**Not run / not claimed:** GPU E2E, FA3 numerical parity, throughput benchmarks, or production model packing. Those belong with each model adapter implementation. No dependency pins or GPU CI registration are changed here.

### API and Usage Example

The existing `(architecture, algorithm)` training-adapter registry remains the dispatch mechanism. An implementing adapter must:

1. Declare `supports_remove_padding = True` only when its packed and non-packed paths are complete.
2. Implement `apply_remove_padding(module, model_config)` to install packed execution without changing parameter names.
3. Branch its existing prepare/forward/output logic on `model_config.use_remove_padding`, including old-policy/reference inference.

The common engine validates support before loading weights, then calls the hook after custom/AutoModel loading and before LoRA/FSDP wrapping. The initial execution contract is FSDP/FSDP2 with SP=1; VeOmni is rejected explicitly.

Do not enable the flag on an unadapted model: **no built-in adapter opts in in this PR**. Existing recipes retain their default execution.

### Design & Code Changes

- `DiffusionModelConfig` plus Hydra/generated configs expose `use_remove_padding=False`; there is no separate `use_packed_batch` configuration flag.
- `DiffusionModelBase` provides capability validation and the startup hook; engines contain no model-name allowlist.
- `PackedSequenceLayout` supplies independent sample boundaries and equal-Q/K/V self-attention for `[1, total_rows, heads, head_dim]`: FA3 hub v1, padded native SDPA reference, and a standalone Torch varlen utility path. The latter is not a new public attention-backend option.
- Model adapters still own text/video/audio/reference row validity, positions, timesteps and per-sample output restoration. Zero-valued latents are not automatically padding; LLM token shifting, causal attention and next-token logits are not applicable.
- Dynamic batch sizing, DP scheduling, rollout batching, loss weighting and existing batch divisibility are unchanged.

Self-review: **READY for the interface-only Draft scope**. No blocking finding in the full 12-file diff. Shared utilities intentionally await model consumers; the absence of an enabled built-in adapter and GPU validation is explicit rather than presented as completed model support.

### Checklist Before Submitting

- [x] Read the contribution guide and applicable config, pipeline, code-style and testing rules.
- [x] Apply pre-commit checks across all tracked files (host-Git workaround documented above).
- [x] Add / update documentation.
- [x] Add CPU tests using the existing `*_on_cpu.py` discovery convention; no workflow changes needed.

### AI assistance and human accountability

AI assistance (OpenAI Codex) was used for implementation, test execution and this description. The human submitter confirmed line-by-line review and relevant test execution during the authoring session, owns this contribution, and is responsible for the final split diff and explaining the change end-to-end. This remains a Draft, without a GPU CI label or a claim of maintainer approval.

### CODE REVIEW

**Comment by zhtmike** [comment]
@NancyFyong per offline discussion, let us see if we can support diffusion model with sequence packing native support in veomni side. 

cc @FoolPlayer @SamitHuang

**Comment by NancyFyong** [comment]
> @NancyFyong per offline discussion, let us see if we can support diffusion model with sequence packing native support in veomni side.
> 
> cc @FoolPlayer @SamitHuang

agree, i think it should be supported in veomni, and we should update verl-omni's veomni side to lora.
